// truffle-decoder/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const truffle_codec_utils_1 = require("truffle-codec-utils");
function getContractNode(contract) {
    return (contract.ast || { nodes: [] }).nodes.find((contractNode) => contractNode.nodeType === "ContractDefinition"
        && (contractNode.name === contract.contractName
            || contractNode.name === contract.contract_name));
}
exports.getContractNode = getContractNode;
function makeContext(contract, node, isConstructor = false) {
    let abi = truffle_codec_utils_1.AbiUtils.schemaAbiToAbi(contract.abi);
    return {
        contractName: contract.contractName,
        binary: isConstructor ? contract.bytecode : contract.deployedBytecode,
        contractId: node.id,
        contractKind: node.contractKind,
        isConstructor,
        abi: truffle_codec_utils_1.AbiUtils.computeSelectors(abi),
        payable: truffle_codec_utils_1.AbiUtils.abiHasPayableFallback(abi),
        compiler: contract.compiler
    };
}
exports.makeContext = makeContext;
